### PR TITLE
Feature/csl880 hd discovery

### PR DESCRIPTION
**Pull request by pva701**
Implemented HD address discovery

### CODE REVIEW

File: core/Pos/Crypto/HD.hs [hunk 1]
```diff
@@ -1,30 +1,35 @@
...
 import           Universum
 
-import           Pos.Binary.Class             (encodeStrict)
+import           Pos.Binary.Class             (decodeFull, encodeStrict)
 import           Pos.Crypto.Signing           (PublicKey (..), SecretKey (..))
 
 -- | Passphrase is a hash of root public key.
--- We don't use root public key to store money, we use hash of it instead.
-newtype HDPassphrase = HDPassphrase ByteString
+-- We don't use root public key for store money, we use hash of it instead.
```
Comment (Martoon-00) [inline core/Pos/Crypto/HD.hs]: Keh, missed during merge?

Comment (pva701) [inline core/Pos/Crypto/HD.hs]: yep, sorry

Comment (pva701) [inline core/Pos/Crypto/HD.hs]: don't say to gromak plz, he'll kill me

File: core/Pos/Crypto/HD.hs [hunk 2]
```diff
@@ -77,25 +84,46 @@ deriveHDSecretKey passPhrase (SecretKey xprv) childIndex
...
+            Left er ->
+                fail $ "Error in unpackHDAddressAttr, during deserialization: " <> show er
+            Right path -> pure path
+
 -- Wrapper around ChaChaPoly1305 module.
 encryptChaChaPoly
     :: ByteString -- nonce (12 random bytes)
-    -> ByteString -- symmetric key
+    -> ByteString -- symmetric key (must be 32 bytes)
     -> ByteString -- optional associated data (won't be encrypted)
```
Comment (georgeee) [inline core/Pos/Crypto/HD.hs]: Please add comment that header is chunk of data we want to transfer unecncrypted but still want it to be part of tag digest (so tag verifies validity of both encrypted data and unencrypted header)

File: core/Pos/Crypto/HD.hs (hunk 1 above)
Comment (pva701) [inline core/Pos/Crypto/HD.hs]: @georgeee noooooooooooooooooooooo, you betrayed me...

Comment (Martoon-00) [inline core/Pos/Crypto/HD.hs]: Looool :rofl:

File: core/Pos/Crypto/HD.hs [hunk 3]
```diff
@@ -77,25 +84,46 @@ deriveHDSecretKey passPhrase (SecretKey xprv) childIndex
...
+    let !unpackCF =
+          decryptChaChaPoly
+              addrAttrNonce
+              passphrase
+              addrAttrHeader
+              payload
+    case unpackCF of
+        Left er ->
+            fail $ "Error in unpackHDAddressAttr, during decryption: " <> show er
+        Right p -> case decodeFull . BSL.fromStrict $ p of
```
Comment (pva701) [inline core/Pos/Crypto/HD.hs]: ok

File: core/Pos/Crypto/HD.hs (hunk 2 above)
Comment (pva701) [inline core/Pos/Crypto/HD.hs]: mmm, nice. thank you for this clarify information.

File: core/Pos/Crypto/HD.hs (hunk 3 above)
Comment (georgeee) [inline core/Pos/Crypto/HD.hs]: PVA-drivven development

File: core/Pos/Crypto/HD.hs [hunk 4]
```diff
@@ -124,7 +128,7 @@ unpackHDAddressAttr (HDPassphrase passphrase) (HDAddressPayload payload) = do
 encryptChaChaPoly
     :: ByteString -- nonce (12 random bytes)
     -> ByteString -- symmetric key (must be 32 bytes)
-    -> ByteString -- optional associated data (won't be encrypted)
+    -> ByteString -- optional associated data (for more information see @addrAttrHeader@)
```
Comment (georgeee) [inline core/Pos/Crypto/HD.hs]: Nooo, please insert info directly here and get rid of `addrAttrHeader` completely

File: core/Pos/Crypto/HD.hs [hunk 5]
```diff
@@ -105,3 +133,29 @@ encryptChaChaPoly nonce key header plaintext = do
...
+
+toEither :: CryptoFailable a -> Either Text a
+toEither (CryptoPassed x)  = pure x
+toEither (CryptoFailed er) = Left $ show er
+
+-- Wrapper around ChaChaPoly1305 module.
+decryptChaChaPoly
+    :: ByteString -- nonce (12 random bytes)
+    -> ByteString -- symmetric key
+    -> ByteString -- optional associated data (for more information see @addrAttrHeader@)
```
Comment (georgeee) [inline core/Pos/Crypto/HD.hs]: @pva701 no `addrAttrHeader` here

File: core/Pos/Crypto/HD.hs [hunk 6]
```diff
@@ -77,31 +84,78 @@ deriveHDSecretKey passPhrase (SecretKey xprv) childIndex
...
+              addrAttrNonce
+              passphrase
+              ""
               pathSer
     case packCF of
-        CryptoFailed er -> error $ "Error during packHDAddressAttr: " <> show er
+        CryptoFailed er -> error $ "Error in packHDAddressAttr: " <> show er
         CryptoPassed p  -> HDAddressPayload p
 
+unpackHDAddressAttr :: MonadFail m => HDPassphrase -> HDAddressPayload -> m [Word32]
```
Comment (Martoon-00) [inline core/Pos/Crypto/HD.hs]: Opportunity to generate `HDAddressPayload` within CSL-879, with which nothing can be done then, makes me :smile:

**Comment by rockbmb** [comment]
@Martoon-00 Is this good to go?

**Comment by Martoon-00** [comment]
@rockbmb Yep, everything seems fine to me here

**Pull request merged**